src/test/scala/org/broadinstitute/dsde/workbench/leonardo/dao/google/MockGoogleDataprocDAO.scala

```diff
@@ -93,6 +97,22 @@ class MockGoogleDataprocDAO(ok: Boolean = true) extends GoogleDataprocDAO {
   }
 
   override def resizeCluster(googleProject: GoogleProject, clusterName: ClusterName, numWorkers: Option[Int], numPreemptibles: Option[Int]): Future[Unit] = {
```

Comment:
Should we do any validation on the value of `numWorkers`? What happens if it's a negative number for instance?

Reply:
yeah that would be a good idea. there's some strange criteria for the number of workers allowed so i should capture that in a validation step so we can fail fast

Reply:
i played around with this a bit and i think im going to leave the validation to google so we don't duplicate their logic. they're able to fail fast if an invalid number of workers is specified so it's essentially the same UX

Reply:
the call to edit the machine config of the cluster in google isn't asynchronous, it happens and completes when the user calls the PATCH endpoint. the async part is the actual provisioning/decommissioning of the worker nodes. because google does the validation upon editing the machine config, we can fail fast and immediately return their error message to the user. their logic isn't totally straight forward which is another reason why im hesitant to reimplement it in leo; google knows best what their restrictions are. i.e. a cluster with only a master node can't be resized to have more workers, a cluster with workers can't be resized down to 0, there must be 2 or more worker nodes, there can be any number > 0 of preemptible worker nodes, etc.

the actual message itself is user-friendly and says exactly what the problem is, but it's buried underneath a giant stacktrace. i think a good middle-ground here would be to rip the message out of the GoogleJsonResponseException and throw that into a new type of LeoException. that way we still let Google do the work of validating but we return something user-friendly

for reference, here's what that error message looks like. the message itself can vary but here's an example:

```
{
  "source": "leonardo",
  "causes": [],
  "exceptionClass": "com.google.api.client.googleapis.json.GoogleJsonResponseException",
  "stackTrace": [
    {
      "className": "com.google.api.client.googleapis.json.GoogleJsonResponseException",
      "methodName": "from",
      "fileName": "GoogleJsonResponseException.java",
      "lineNumber": 146
    },
    {
      "className": "com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest",
      "methodName": "newExceptionOnError",
      "fileName": "AbstractGoogleJsonClientRequest.java",
      "lineNumber": 113
    },
    {
      "className": "com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest",
      "methodName": "newExceptionOnError",
      "fileName": "AbstractGoogleJsonClientRequest.java",
      "lineNumber": 40
    },
    {
      "className": "com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1",
      "methodName": "interceptResponse",
      "fileName": "AbstractGoogleClientRequest.java",
      "lineNumber": 321
    },
    {
      "className": "com.google.api.client.http.HttpRequest",
      "methodName": "execute",
      "fileName": "HttpRequest.java",
      "lineNumber": 1065
    },
    {
      "className": "com.google.api.client.googleapis.services.AbstractGoogleClientRequest",
      "methodName": "executeUnparsed",
      "fileName": "AbstractGoogleClientRequest.java",
      "lineNumber": 419
    },
    {
      "className": "com.google.api.client.googleapis.services.AbstractGoogleClientRequest",
      "methodName": "executeUnparsed",
      "fileName": "AbstractGoogleClientRequest.java",
      "lineNumber": 352
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.GoogleUtilities",
      "methodName": "$anonfun$executeGoogleCall$1",
      "fileName": "GoogleUtilities.scala",
      "lineNumber": 70
    },
    {
      "className": "scala.util.Try$",
      "methodName": "apply",
      "fileName": "Try.scala",
      "lineNumber": 209
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.GoogleUtilities",
      "methodName": "executeGoogleCall",
      "fileName": "GoogleUtilities.scala",
      "lineNumber": 70
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.GoogleUtilities",
      "methodName": "executeGoogleCall$",
      "fileName": "GoogleUtilities.scala",
      "lineNumber": 67
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.AbstractHttpGoogleDAO",
      "methodName": "executeGoogleCall",
      "fileName": "AbstractHttpGoogleDAO.scala",
      "lineNumber": 14
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.GoogleUtilities",
      "methodName": "executeGoogleRequest",
      "fileName": "GoogleUtilities.scala",
      "lineNumber": 51
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.GoogleUtilities",
      "methodName": "executeGoogleRequest$",
      "fileName": "GoogleUtilities.scala",
      "lineNumber": 50
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.AbstractHttpGoogleDAO",
      "methodName": "executeGoogleRequest",
      "fileName": "AbstractHttpGoogleDAO.scala",
      "lineNumber": 14
    },
    {
      "className": "org.broadinstitute.dsde.workbench.leonardo.dao.google.HttpGoogleDataprocDAO",
      "methodName": "$anonfun$resizeCluster$2",
      "fileName": "HttpGoogleDataprocDAO.scala",
      "lineNumber": 197
    },
    {
      "className": "akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread$$anon$3",
      "methodName": "block",
      "fileName": "ThreadPoolBuilder.scala",
      "lineNumber": 167
    },
    {
      "className": "akka.dispatch.forkjoin.ForkJoinPool",
      "methodName": "managedBlock",
      "fileName": "ForkJoinPool.java",
      "lineNumber": 3641
    },
    {
      "className": "akka.dispatch.MonitorableThreadFactory$AkkaForkJoinWorkerThread",
      "methodName": "blockOn",
      "fileName": "ThreadPoolBuilder.scala",
      "lineNumber": 165
    },
    {
      "className": "akka.dispatch.BatchingExecutor$BlockableBatch",
      "methodName": "blockOn",
      "fileName": "BatchingExecutor.scala",
      "lineNumber": 106
    },
    {
      "className": "scala.concurrent.package$",
      "methodName": "blocking",
      "fileName": "package.scala",
      "lineNumber": 142
    },
    {
      "className": "org.broadinstitute.dsde.workbench.google.GoogleUtilities",
      "methodName": "$anonfun$retryWhen500orGoogleError$3",
      "fileName": "GoogleUtilities.scala",
      "lineNumber": 42
    },
    {
      "className": "scala.concurrent.Future$",
      "methodName": "$anonfun$apply$1",
      "fileName": "Future.scala",
      "lineNumber": 654
    },
    {
      "className": "scala.util.Success",
      "methodName": "$anonfun$map$1",
      "fileName": "Try.scala",
      "lineNumber": 251
    },
    {
      "className": "scala.util.Success",
      "methodName": "map",
      "fileName": "Try.scala",
      "lineNumber": 209
    },
    {
      "className": "scala.concurrent.Future",
      "methodName": "$anonfun$map$1",
      "fileName": "Future.scala",
      "lineNumber": 288
    },
    {
      "className": "scala.concurrent.impl.Promise",
      "methodName": "liftedTree1$1",
      "fileName": "Promise.scala",
      "lineNumber": 29
    },
    {
      "className": "scala.concurrent.impl.Promise",
      "methodName": "$anonfun$transform$1",
      "fileName": "Promise.scala",
      "lineNumber": 29
    },
    {
      "className": "scala.concurrent.impl.CallbackRunnable",
      "methodName": "run",
      "fileName": "Promise.scala",
      "lineNumber": 60
    },
    {
      "className": "akka.dispatch.BatchingExecutor$AbstractBatch",
      "methodName": "processBatch",
      "fileName": "BatchingExecutor.scala",
      "lineNumber": 55
    },
    {
      "className": "akka.dispatch.BatchingExecutor$BlockableBatch",
      "methodName": "$anonfun$run$1",
      "fileName": "BatchingExecutor.scala",
      "lineNumber": 91
    },
    {
      "className": "scala.runtime.java8.JFunction0$mcV$sp",
      "methodName": "apply",
      "fileName": "JFunction0$mcV$sp.java",
      "lineNumber": 12
    },
    {
      "className": "scala.concurrent.BlockContext$",
      "methodName": "withBlockContext",
      "fileName": "BlockContext.scala",
      "lineNumber": 81
    },
    {
      "className": "akka.dispatch.BatchingExecutor$BlockableBatch",
      "methodName": "run",
      "fileName": "BatchingExecutor.scala",
      "lineNumber": 91
    },
    {
      "className": "akka.dispatch.TaskInvocation",
      "methodName": "run",
      "fileName": "AbstractDispatcher.scala",
      "lineNumber": 40
    },
    {
      "className": "akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask",
      "methodName": "exec",
      "fileName": "ForkJoinExecutorConfigurator.scala",
      "lineNumber": 44
    },
    {
      "className": "akka.dispatch.forkjoin.ForkJoinTask",
      "methodName": "doExec",
      "fileName": "ForkJoinTask.java",
      "lineNumber": 260
    },
    {
      "className": "akka.dispatch.forkjoin.ForkJoinPool$WorkQueue",
      "methodName": "runTask",
      "fileName": "ForkJoinPool.java",
      "lineNumber": 1339
    },
    {
      "className": "akka.dispatch.forkjoin.ForkJoinPool",
      "methodName": "runWorker",
      "fileName": "ForkJoinPool.java",
      "lineNumber": 1979
    },
    {
      "className": "akka.dispatch.forkjoin.ForkJoinWorkerThread",
      "methodName": "run",
      "fileName": "ForkJoinWorkerThread.java",
      "lineNumber": 107
    }
  ],
  "message": "400 Bad Request\n{\n  \"code\" : 400,\n  \"errors\" : [ {\n    \"domain\" : \"global\",\n    \"message\" : \"Cluster configuration must have at least 2 workers.\",\n    \"reason\" : \"badRequest\"\n  } ],\n  \"message\" : \"Cluster configuration must have at least 2 workers.\",\n  \"status\" : \"INVALID_ARGUMENT\"\n}"
}```

Reply:
I see. The middle-ground plan sounds great then.